src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs

```diff
@@ -261,52 +279,54 @@ public static void AddProviderTransparent(TypeDescriptionProvider provider, obje
         /// </summary>
         private static void CheckDefaultProvider(Type type)
         {
-            if (s_defaultProviders.ContainsKey(type))
+            s_processedTypesLock.EnterUpgradeableReadLock();
+            try
             {
-                return;
-            }
-
-            lock (s_internalSyncObject)
-            {
-                if (s_defaultProviders.ContainsKey(type))
+                if (!s_processedTypes.Contains(type))
                 {
-                    return;
-                }
-
-                // Immediately clear this. If we find a default provider
-                // and it starts messing around with type information,
-                // this could infinitely recurse.
-                s_defaultProviders[type] = null;
-            }
+                    s_processedTypesLock.EnterWriteLock();
+                    try
+                    {
+                        s_processedTypes.Add(type);
+
+                        // Always use core reflection when checking for
+                        // the default provider attribute. If there is a
+                        // provider, we probably don't want to build up our
+                        // own cache state against the type. There shouldn't be
+                        // more than one of these, but walk anyway. Walk in
+                        // reverse order so that the most derived takes precidence.
+                        var attrs = type.GetCustomAttributes<TypeDescriptionProviderAttribute>(false);
+                        bool providerAdded = false;
+                        foreach(var currentAttr in attrs)
```

Comment:
Why was this going in reverse order previously?

Also, unless something changed recently, a `for` loop indexing over an array is faster than `foreach` since it doesn't require an enumerator and calls to it.

Reply:
@steveharter , thank's for your review. Am I correctly understand, that perfomance and code of [bugfix/30024-problem-with-concurrency-typeDescriptor-getconverter(NEW)](https://github.com/Maximys/runtime/tree/bugfix/30024-problem-with-concurrency-typeDescriptor-getconverter(NEW)) branch is acceptable and I only should change foreach loop to for and apply reverse order (it's my mistake, I have idea of test for prevent this mistake for future)?

Reply:
Why a new branch? I took a quick look at that - can we not using the Linq namespace (which means we need to enumerate the attributes differently) since pulling in Linq is expensive.

Reply:
Linq needs for using [ToArray()](https://github.com/Maximys/runtime/blob/8d214e352c803acdf06a04a6b7fb39336d10d46d/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs#L302) method

Reply:
On Linq - sorry I thought the original didn't have it.

Please apply that other branch here. We need to get this in ASAP since we're locking down for this release. Thanks

Reply:
Ready, but there are some failed tests. I think, they are not linked with my changes, but I'm not sure. Please take a look